typed-racket-lib/typed-racket/private/type-contract.rkt

```diff
@@ -529,8 +529,16 @@
        [(Mutable-VectorTop:)
         (only-untyped mutable-vector?/sc)]
        [(Box: t) (box/sc (t->sc/both t))]
-       [(Pair: t1 t2)
-        (cons/sc (t->sc t1) (t->sc t2))]
+       [(Pair: _ _)
+        (match type
+         [(List: elem-t*)
+          (apply list/sc (map t->sc elem-t*))]
+         [(List: elem-t* #:tail rest-t*)
```

Comment:
This seems like it has quadratic complexity in the length of the list for improper lists.